### ISSUE TITLE
The persistent_id of a ComponentDefinition always returns zero (0)

### ISSUE THREAD

**Issue by 32edges** [SketchUp/api-issue-tracker#29]
[TestModel.zip](https://github.com/SketchUp/api-issue-tracker/files/1660241/TestModel.zip)
SU2017 Pro / SU2018 Pro
Windows 7 

The persistent_id of a ComponentDefinition always return zero (0).  The following entered into the Ruby Console with three component definitions shows that they do respond to 'persistent_id' but always return zero.
```
Sketchup.active_model.definitions.map { |d| [d.respond_to?('persistent_id'), d.persistent_id] }
```
responds with

```
 [[true, 0], [true, 0], [true, 0]]
```

**Comment by thomthom** [comment]
I can reproduce. I have to check if the documentation is wrong, or whether this is a bug.

**Comment by thomthom** [comment]
I looked up the source and found that the documentation isn't accurate. The items you get PIDs from are:

* Entities that you can iterate over in a `Sketchup::Entities` collection.
* `Sketchup::Curve`
* `Sketchup::Page`
* `Sketchup::Vertex`

Persistent ids are intended for things you find in the model - meaning component instances, but not definitions.

May I ask what you tried to use PIDs for component definitions for?

**Comment by DanRathbun** [comment]
> Persistent ids are intended for things you find in the model - meaning component instances, but not definitions.

Oh duh! Makes complete sense (now that you explicitly say it.)

Ie, ALL component definitions are unique anyway, but "The guid changes after the component definition is modified and the component edit is exited."

I would think the OP and others may wish to identify definitions from one session to the next, and the **`#guid`** could do this if "collected" just before the session ends (model is saved and closed.)

But during the session, when a definition is edited, the changing `#guid` is a bit annoying, but the `#object_id` does **not** change.

**Comment by DanRathbun** [comment]
> @thomthom I looked up the source and found that the documentation isn't accurate. 

Surprise, not ... :roll_eyes: 

> @thomthom The items you get PIDs from are: ...

Well hindsight here. **This feature should not have been implemented using inheritance.**

There are just way too many subclasses of **`Sketchup::Entity`** such as all the collection classes, hash classes **(`Sketchup::RenderingOptions`** and **`Sketchup::ShadowInfo`**) which are not part of entities, or property classes like **`Sketchup::Behavior`**.

One might say that the **`Sketchup::Drawingelement`** class would have been a better choice into which to define the **`#persistent_id`** method. (But then you've since added the functionality to pages which are not a drawing element. I also predict layers, materials and perhaps cameras getting this feature in the near future.)

-----

This was (is still) a perfect candidate feature to implement (re-implement) using **composition** and a **mixin module**. Call it something like **`Sketchup::PersistentIdentification`** and only mix it into those classes that will have this functionality.

Mixing into superclasses makes the mixin module an ancestor class for subclasses. So for example mixed into **`Sketchup::Curve`** the mixin module ancestry is inherited by **`Sketchup::ArcCurve`**. The same for mixing into **`Sketchup:;Dimension`**, ... the several dimension subclasses will get the function as well.

?

**Comment by thomthom** [comment]
> I would think the OP and others may wish to identify definitions from one session to the next, and the #guid could do this if "collected" just before the session ends (model is saved and closed.)

I think the definition guid changes as the definition changes. (It's different from the instance guid ... yea, messy inconsistency, I know.)

**Comment by DanRathbun** [comment]
>@thomthom I think the definition guid changes as the definition changes.

Yes, I noted that above. ie ...

> @DanRathbun (So a workaround of sorts is to use the `#guid` from one session to the next, and `#object_id` during the edit session.)

**Comment by Eneroth3** [comment]
> I think the definition guid changes as the definition changes. (It's different from the instance guid ... yea, messy inconsistency, I know.)

The definition GUID _must_ change on each definition change. The GUID is how SketchUp distinguishes components and knows when a newly loaded/pasted component is identical to one already present in the model or is a separate component.

> One might say that the Sketchup::Drawingelement class would have been a better choice into which to define the #persistent_id method

And one might even say that a ComponentDefinition shouldn't be a DrawingElement as it's not something you can select in the viewport, just something that is referenced from such an entity, a bit like a material or layer.

**Comment by DanRathbun** [comment]
> @Eneroth3 And one might even say that a ComponentDefinition shouldn't be a DrawingElement as it's not something you can select in the viewport, just something that is referenced from such an entity, a bit like a material or layer.

True! And there are some methods that are inherited from **`Drawingelement`** class, that are weird for a definition, that we'd expect to be in the **`Behavior`** class. But the ancestry (and documentation quirks) of the **`Sketchup::ComponentDefinition`** class is a separate issue (that could deserve it's own tracker issue.)

-----

@thomthom 
How hard is it on the C-side to implement Ruby composition (mixin module) ?
If it is not impossible, please file a "enhancement" for reimplement as mixin module.

Coders can then do "duck-typing" thus, ...
where the mixin ancestor module is named **`Sketchup::PersistentEntity`** ...
```ruby
if entity.is_a?(Sketchup::PersistentEntity)
  # do code using #persistent_id ...
else
  # do some workaround ...
end
```
This can also be used ...
```ruby
if entity.class.ancestors.include?(Sketchup::PersistentEntity)
```

**Comment by DanRathbun** [comment]
> @thomthom
How hard is it on the C-side to implement Ruby composition (mixin module) ?

I see the answer from the silver hammer docs:
https://silverhammermba.github.io/emberb/c/#other-stuff

```c
rb_include_module( cSketchupDefinition, mSketchupPersistentEntity );
```

Does this "**enhancement**" need to have it's own GitHub issue ?

**Comment by thomthom** [comment]
> Does this "enhancement" need to have it's own GitHub issue ?
I think it would be good to track a separate issue to provide constant modules to be included by the API users. Would be good to make a pass on the whole API and provide such facility.

**Comment by thomthom** [comment]
Logged as SU-38576

**Comment by thomthom** [comment]
This has been fixed and pushed to production a while ago.

**Comment by DanRathbun** [comment]
WHAT was fixed ? We discussed a lot here ...

> The persistent_id of a ComponentDefinition always return zero (0). 

A bug.

... and documentation clarity ...

> I looked up the source and found that the documentation isn't accurate. The items you get PIDs from are:
>
> * Entities that you can iterate over in a Sketchup::Entities collection.
> * Sketchup::Curve
> * Sketchup::Page
> * Sketchup::Vertex
>
>Persistent ids are intended for things you find in the model - meaning component instances, but not definitions.

----

The other stuff I need to open a new issue for if I haven't yet.

**Comment by thomthom** [comment]
The documenation was updated to reflect the actual list of entities that have persistent IDs. However, I thought that had been pushed live already. I see it is not.

If there are types of entities that you wish to see persistent IDs for, please file them as separate feature request. The current set is by design.

**Comment by DanRathbun** [comment]
1. Please reread the thread. The PID functionality was incorrectly implemented via inheritance. It should be re-implemented using composition (a mixin module.)
[... OR ... `Sketchup::ComponentDefinition` should not be a subclass of `Sketchup::DrawingElement` ?]

2. Yes, it appears coders wish that definitions also get persistent PIDs.
